internal/cmd/agent/controllers/bundledeployment/controller_test.go

```diff
@@ -0,0 +1,119 @@
+	fleet "github.com/rancher/fleet/pkg/apis/fleet.cattle.io/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+func TestSetNamespaceLabelsAndAnnotations(t *testing.T) {
+	tests := map[string]struct {
+		bd          *fleet.BundleDeployment
+		ns          corev1.Namespace
+		release     string
+		expectedNs  corev1.Namespace
+		expectedErr error
+	}{
+		"NamespaceLabels and NamespaceAnnotations are appended": {
+			bd: &fleet.BundleDeployment{Spec: fleet.BundleDeploymentSpec{
+				Options: fleet.BundleDeploymentOptions{
+					NamespaceLabels:      &map[string]string{"optLabel1": "optValue1", "optLabel2": "optValue2"},
+					NamespaceAnnotations: &map[string]string{"optAnn1": "optValue1"},
+				},
+			}},
+			ns: corev1.Namespace{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{"name": "test"},
+				},
+			},
+			release: "namespace/foo/bar",
+			expectedNs: corev1.Namespace{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels:      map[string]string{"name": "test", "optLabel1": "optValue1", "optLabel2": "optValue2"},
+					Annotations: map[string]string{"optAnn1": "optValue1"},
+				},
+			},
+			expectedErr: nil,
+		},
+
+		"NamespaceLabels and NamespaceAnnotations removes entries that are not in the options, except the name label": {
+			bd: &fleet.BundleDeployment{Spec: fleet.BundleDeploymentSpec{
+				Options: fleet.BundleDeploymentOptions{
+					NamespaceLabels:      &map[string]string{"optLabel": "optValue"},
+					NamespaceAnnotations: &map[string]string{},
+				},
+			}},
+			ns: corev1.Namespace{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels:      map[string]string{"nsLabel": "nsValue", "name": "test"},
+					Annotations: map[string]string{"nsAnn": "nsValue"},
+				},
+			},
+			release: "namespace/foo/bar",
+			expectedNs: corev1.Namespace{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels:      map[string]string{"optLabel": "optValue", "name": "test"},
+					Annotations: map[string]string{},
+				},
+			},
+			expectedErr: nil,
+		},
+
+		"NamespaceLabels and NamespaceAnnotations updates existing values": {
+			bd: &fleet.BundleDeployment{Spec: fleet.BundleDeploymentSpec{
+				Options: fleet.BundleDeploymentOptions{
+					NamespaceLabels:      &map[string]string{"bdLabel": "labelUpdated"},
+					NamespaceAnnotations: &map[string]string{"bdAnn": "annUpdated"},
+				},
+			}},
+			ns: corev1.Namespace{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels:      map[string]string{"bdLabel": "nsValue"},
+					Annotations: map[string]string{"bdAnn": "nsValue"},
+				},
+			},
+			release: "namespace/foo/bar",
+			expectedNs: corev1.Namespace{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels:      map[string]string{"bdLabel": "labelUpdated"},
+					Annotations: map[string]string{"bdAnn": "annUpdated"},
+				},
+			},
+			expectedErr: nil,
+		},
```

Comment:
Would it make sense to add a test case where the namespace is not found?

Reply:
I had to create different mocks for that. I didn't do it because I thought it was not worh it and it was ok with the other test cases. However you might be right, so I added it in https://github.com/rancher/fleet/pull/1627/commits/d06724c7cba0b85ab6a124f3064bd1692dd98572